Stream interface, more flags and error handling

sbencoding opened the pull request:
Fixes #8
  
🦄 => I've read the message
- [x] Implement `/apps` flag - done with `apps: true` option
- [x] Implement `/svc` flag - done with `services: true` option
- [x] Implement `/m` flag - done with `modules: 'example.dll'` option
- [x] Implement a Stream interface (see prototype: [tasklist-stream](https://github.com/MarkTiedemann/tasklist-stream)) - done based on prototype, + added checking for empty results
- [x] Implement a check and/or add appropriate documentation that, when querying remote machines, `windowTitle` and `status` filters are not supported (see: `tasklist /?`) - done, written in the readme and application throws error if these parameters are specified with the remote machine option

I also wrote tests for the new flags, and updated the old ones, since they didn't work anymore with the new stream interface


<!-- Issuehunt content -->

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Referenced issues

This pull request has been submitted to:
- [#8: Open todos](https://issuehunt.io/repos/36579963/issues/8)
---

IssueHunt has been backed by the following sponsors. [Become a sponsor](https://issuehunt.io/membership/members)
</details>
<!-- /Issuehunt content-->

sbencoding committed:
Implement Stream interface
Add apps, modules, services flag
Update test and documentation
Handle wrong option combinations

sbencoding committed:
Pass no-unused-vars test

sbencoding committed:
Fix travisCI failing app tests

MarkTiedemann commented:
I believe the `Promise` interface should stay intact. The `Stream` interface was supposed to be an "add-on", for example, exposed as `tasklist.stream`. If implemented that way, this PR would be a minor change, rather than a major breaking change.

PS: The streaming interface is intended for use cases in which you're not interested in all processes / services / apps, but only in a limited number. So you don't need to wait for all results, but can abort once the item(s) that you were looking for has/have been received. This is great for performance as well since `tasklist` can be quite slow on some systems.

sbencoding commented:
Oops, my bad.  
Working on it right now...

MarkTiedemann commented:
> Oops, my bad.

No, it is my bad. I should have given more details in the issue that you're now fixing.

Great job, by the way! :)

sbencoding committed:
Add back promise interface
Write tests for promise interface
Update readme with promise interface

sbencoding committed:
Better jsdoc strings with typedef

sbencoding committed:
Ignore incomplete csv chunk parse error

sbencoding committed:
Close returned stream in case parser throws error

sbencoding commented:
Ok, I've added the `promise` interface back.  
As you've suggested the `stream` interface can be accessed by: `tasklist.stream()`, and the `promise` interface by: `tasklist()`.  
I've also updated the documentation and the tests, and finally the TravisCI build allso passes 😄

MarkTiedemann commented on index.js:
```diff
@@ -31,38 +95,91 @@ module.exports = async (options = {}) => {
+ */
+async function promiseInterface(options = {}) {
+	const {args, columns, currentTransform} = main(options);
+	const {stdout} = await execFile('tasklist', args);
```
```suggestion
	const {stdout} = await execFile('tasklist.exe', args);
```

Like in L127.

MarkTiedemann commented on index.js:
```diff
@@ -31,38 +95,91 @@ module.exports = async (options = {}) => {
+ * @param {Options} options The options of the command
+ * @returns {Stream} Stream, returning parsed results
+ */
+function stream(options = {}) {
```
You called the other function `promiseInterface`. Perhaps rename to `streamInterface`? Or remove "interface" from both names?

MarkTiedemann commented on index.js:
```diff
@@ -31,38 +95,91 @@ module.exports = async (options = {}) => {
-		task.pid = Number(task.pid);
-		task.sessionNumber = Number(task.sessionNumber);
-		task.memUsage = Number(task.memUsage.replace(/[^\d]/g, '')) * 1024;
+	processOutput.on('data', data => {
```
I'm not sure that this is a good way to implement the streaming. You're manually connecting your streams rather that `.pipe`'ing the data through them. I think that this means that, for example, once the returned stream is aborted, the other streams will continue to handle incoming tasklist data, rather than being aborted, too.

Haven't worked with Node streams in a few years, though, so I might be wrong.

@sindresorhus Any input on this?

MarkTiedemann commented:
> Ok, I've added the promise interface back.

Great, thank you!

> and finally the TravisCI build allso passes

For some reason, GitHub is marking the CI as [`Queued — Build Created`](https://github.com/sindresorhus/tasklist/pull/21/checks?check_run_id=163829845) / yellow to me, whereas Travis shows the build as ["passed"](https://travis-ci.com/sindresorhus/tasklist/builds/118282064) / green ... But yeah, the tests are passing. Also, run some tests on my personal Windows machine, just in case.

I asked for a few minor changes that you could clean-up, @sbencoding. And I'm not sure whether the streaming interface was implemented in a good way - that might be a few more changes.

Other than that, this is looking good to me.

Perhaps @sindresorhus should take a look as well. :)

sbencoding committed:
Implement stream interface using pipes
+ few minor changes

sbencoding committed:
Use multipipe for node v8 support

sbencoding commented:
I've implemented the minor changes, and the `stream` interface now uses `pipes` instead of the manual stream connection.  
In the first version I couldn't find a way to check for empty results while still using the `pipe` *chain*, so that's why I decided to chain it manually, but now I've found out that I can use a `Transform` for that, so it's using `pipes` again.  
`multipipe` is required, becuase the `pipeline` function from nodejs is only supported on v10 and up.  
The module makes sure that every error from every stream in the `pipe` *chain* is handled by a single error handler.  
This is required to catch errors when the user end's the stream, but the one of the streams still tries to write to another that's already ended.  
If there's anything I need change on the new `stream` interface, just let me know :smile:

sindresorhus commented:
> multipipe is required, becuase the pipeline function from nodejs is only supported on v10 and up.

I'm ok with dropping Node.js 8 support. Node.js 8 ends LTS in December, so we can just wait publishing until then.

sindresorhus commented on csv-headers.js:
```diff
@@ -0,0 +1,48 @@
+module.exports = {
+	default: [
```
This file needs to be added to `files` in package.json.

sindresorhus committed:
Update csv-headers.js

sindresorhus commented on index.js:
```diff
@@ -1,23 +1,88 @@
-const sec = require('sec');
+const {promisify} = require('util');
+const pipe = require('multipipe');
+const csv = require('csv');
```
Use `neat-csv`. It support both promise/stream input.

sindresorhus commented on index.js:
```diff
@@ -1,23 +1,88 @@
+ * @property {String} username The username of the remote machine
+ * @property {String} password The password of the remote machine
+ * @property {Array} filter Filters to pass to the command
+ */
```
I don't really want this. It would be better if you could add a TypeScript definition in a follow-up PR, following https://github.com/sindresorhus/typescript-definition-style-guide

sindresorhus commented on readme.md:
```diff
@@ -38,13 +36,42 @@ See the [`tasklist` docs](https://technet.microsoft.com/en-us/library/bb491010.a
+    status: 'Running',
+    username: 'SINDRESORHU3930\\sindre'
+    cpuTime: 0,
+    windowTitle: 'Task Host Window'
```
Use tab-indentation.

sindresorhus commented on test.js:
```diff
@@ -23,9 +23,86 @@ const hasVerboseTaskProps = (t, task) => {
+			reject(error);
+		}
+	});
+};
```
Use https://github.com/sindresorhus/get-stream

sindresorhus commented on test.js:
```diff
@@ -39,13 +116,137 @@ const macro = async (t, options) => {
+
+test('[Stream interface] default', macro, {});
+test('[Stream interface] verbose option', macro, {verbose: true});
+test('[Stream interface] filter option', macro, {filter: ['sessionname eq console', 'username ne F4k3U53RN4M3']});
```
Instead of the title prefix, put stream and promise tests in separate files.

sbencoding committed:
Remove JSDoc comments
Use pipeline instead of multipipe
Add typescript definition file and test

sbencoding committed:
Split tests into multiple files
Test for /svc with /m flag
Use get-stream instead of manual reading of the stream

sbencoding committed:
Install requried packages, remove unused ones
Configure ava to detect multiple test files
Configure xo to ignore .ts files

sbencoding committed:
Merge branch 'master' into master

sbencoding committed:
Fix misimplementation of pipeline

sbencoding committed:
Merge branch 'master' of https://github.com/sbencoding/tasklist

sbencoding committed:
Uncomment, because of testing

sbencoding commented:
I've implemented most of the requested changes, all test pass locally, the TravisCI build only fails for node v8, because as suggested I removed the `multipipe` dependency and used `pipeline` instead, which is built into NodeJS.  
One change that I didn't implement was the `neat-csv` one. Correct me if I'm wrong, but my understanding is that the point of the stream interface is that it returns a stream that can be stopped at any time. While `neat-csv` does accept a stream as it's input it doesn't return a stream, so as I see it, it would defeat the whole purpose of the feature.

Let me know what you think!

sindresorhus commented:
> One change that I didn't implement was the neat-csv one. Correct me if I'm wrong, but my understanding is that the point of the stream interface is that it returns a stream that can be stopped at any time. While neat-csv does accept a stream as it's input it doesn't return a stream, so as I see it, it would defeat the whole purpose of the feature.

Yeah, never mind that.

sindresorhus committed:
Update index.d.ts

sindresorhus commented:
The `index.d.ts` should be added in a separate PR. This PR already large enough. Also make sure you then adhere to: https://github.com/sindresorhus/typescript-definition-style-guide

sindresorhus commented on global.test.js:
```diff
@@ -0,0 +1,40 @@
+			password: 'test',
+			filter: ['Windowtitle eq asd']
+		});
+	});
```
`throwsAsync` accepts a promise.

```js
await t.throwsAsync(
	tasklist({
		system: 'test',
		username: 'test',
		password: 'test',
		filter: ['Windowtitle eq asd']
	})
);
```

sindresorhus commented on global.test.js:
```diff
@@ -0,0 +1,40 @@
+import test from 'ava';
```
Can you move all the test files to a `test` directory?

sindresorhus commented on index.js:
```diff
@@ -1,23 +1,66 @@
+const {pipeline} = require('stream');
+const csv = require('csv');
+const csvHeaders = require('./csv-headers');
+const transform = require('./transform');
```
The `transform` file needs to be added to package.json

sindresorhus commented on index.js:
```diff
@@ -1,23 +1,66 @@
-module.exports = async (options = {}) => {
+csv.parse[promisify.custom] = (input, options) => new Promise(resolve => {
+	csv.parse(input, options, (_, records) => resolve(records));
+});
```
Why can't you just use `promisify` directly? And if you're going to do this, you might as well just drop using `promisify`.

sindresorhus commented on readme.md:
````diff
@@ -38,13 +36,42 @@ See the [`tasklist` docs](https://technet.microsoft.com/en-us/library/bb491010.a
+
+```js
+const tasklist = require('tasklist');
+tasklist.stream({verbose: true}).on('data', (process) => console.log(process))
````
```suggestion
tasklist.stream({verbose: true}).pipe(process.stdout);
```

would be a better example.

sindresorhus commented on readme.md:
```diff
@@ -38,13 +36,42 @@ See the [`tasklist` docs](https://technet.microsoft.com/en-us/library/bb491010.a
+The `system`, `username`, `password` options must be specified together.  
+The `modules` and `services` options can't be specified if verbose is set to `true`.  
+The `modules` and `services` options can't be specified at the same time.  
+When `system`, `username`, `password` options are specified, the filter option can't have `windowtitle` and `status` as the parameter.  
```
Are all of these conditions enforced in code? If not, they should be.

sindresorhus commented:
This is looking very good.

sbencoding committed:
Update examples

sbencoding committed:
Setup test folder
Validate remote system parameters

sbencoding committed:
Remove ts defs and tests for future PR

sbencoding committed:
Pass directly instead of wrapping in function

sbencoding committed:
merge from remote

sbencoding committed:
Fix import order

sbencoding commented:
Ok, I implemented the requested changes, TravisCI only fails the node version 8 build/tests.  
I removed the `index.d.ts` and `index.test-d.ts` files as well as the `tsd-check` from the test script to be readded in a future PR as requested.

sindresorhus commented:
Thanks for working on this 🙌

> I removed the index.d.ts and index.test-d.ts files as well as the tsd-check from the test script to be readded in a future PR as requested.

Do you plan to work on this soon? I'm asking because I need to know whether I should do a new release or wait.

sbencoding commented:
Thanks for merging!
I already have those files ready (sort of). I'm just going to re-check everything and also check the style guide you've sent and probably submit it in a PR tomorrow.
